File: src/test/java/com/twsg/lnl/kotlin/c5_delegate/map/g1_java/Grade.java
package com.twsg.lnl.kotlin.c5_delegate.map.g1_java;

import java.util.Map;

/**
 * @author tonyl
 */
public class Grade {

    private final Map<String, Double> gradeMap;

    public Grade(Map<String, Double> gradeMap) {
        this.gradeMap = gradeMap;
    }

    public double getMath() {
        return gradeMap.computeIfAbsent("math", key -> Double.NaN);
    }

    public void setMath(double grade) {
        gradeMap.put("math", grade);
    }

    public double getPhysics() {
        return gradeMap.computeIfAbsent("physics", key -> Double.NaN);
    }

    public void setPhysics(double grade) {
        gradeMap.put("physics", grade);
    }
}

File: src/test/java/com/twsg/lnl/kotlin/c5_delegate/map/g2_kotlin/Grade.kt
package com.twsg.lnl.kotlin.c5_delegate.map.g2_kotlin

/**
 * @author tonyl
 */

class Grade(private val gradeMap: MutableMap<String, Double>) {

    /* These properties are backed by a Map via delegates */

    var math by gradeMap

    var physics by gradeMap

}